### PR TITLE
river: init at unstable-2021-04-08

### PR DESCRIPTION
**Pull request by heavyrain266**
###### Motivation for this change
Hi, I just wanted to add new Wayland Compositor into master branch, it's quite popular but just saw that a lot of people have problems with building/installing it on nixos. I have tested this derivation by building it through selhosted hydra, locally on nixos with nix-build and on non-nixos distribution locally and from own channel. I'm not sure about versioning in commit message since they doesn't provide any github releases yet.

Here is github page for this package https://github.com/ifreund/river

###### Things done 

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by nixinator** [comment]
is there anyway you get them to make a tagged or release, sometimes if you ask nicely...they will :-).

**Comment by branwright1** [comment]
> is there anyway you get them to make a tag/release, sometimes if you ask nicely...they will :-).

I will ask them later today, haven't thinked about it.

**Comment by fortuneteller2k** [comment]
> is there anyway you get them to make a tag/release, sometimes if you ask nicely...they will :-).

If this isn't possible, just make the version `unstable-yyyy-mm-dd` with the date of the commit.

File: pkgs/applications/window-managers/river/default.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  preBuild = "export HOME=$TMPDIR;";
+  installPhase = "zig build -Drelease-safe -Dxwayland -Dman-pages --prefix $out install";
+
+  nativeBuildInputs = [ zig scdoc pkgconfig ];
+
+  installFlags = [ "DESTDIR=$(out)" ];
+
+  meta = with lib; {
+    description = "A dynamic tiling wayland compositor";
+    longDescriotion = ''
```
Comment (fortuneteller2k) [inline pkgs/applications/window-managers/river/default.nix:31]:
```suggestion
    longDescription = ''
```

File: pkgs/applications/window-managers/river/default.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  };
+
+  buildInputs = [ wayland xwayland wayland-protocols wlroots pixman
+    libxkbcommon pixman libudev libevdev libX11 libGL
+  ];
+
+  preBuild = "export HOME=$TMPDIR;";
+  installPhase = "zig build -Drelease-safe -Dxwayland -Dman-pages --prefix $out install";
+
+  nativeBuildInputs = [ zig scdoc pkgconfig ];
```
Comment (fortuneteller2k) [inline pkgs/applications/window-managers/river/default.nix:25]:
```suggestion
  nativeBuildInputs = [ zig scdoc pkg-config ];
```

File: pkgs/applications/window-managers/river/default.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
+{ lib, stdenv ,fetchFromGitHub
+, zig, pkgconfig, scdoc
```
Comment (fortuneteller2k) [inline pkgs/applications/window-managers/river/default.nix:2]:
```suggestion
, zig, pkg-config, scdoc
```

**Review by moni-dz — CHANGES_REQUESTED** [review]

File: pkgs/applications/window-managers/river/default.nix (hunk 2 above)
Comment (fortuneteller2k) [reply pkgs/applications/window-managers/river/default.nix:25]: `pkgconfig` is an alias for `pkg-config`

File: pkgs/applications/window-managers/river/default.nix (hunk 3 above)
Comment (fortuneteller2k) [reply pkgs/applications/window-managers/river/default.nix:2]: `pkgconfig` is an alias for `pkg-config`

File: pkgs/applications/window-managers/river/default.nix [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+  meta = with lib; {
+    description = "A dynamic tiling wayland compositor";
+    longDescriotion = ''
+      river is a dynamic tiling wayland compositor that takes inspiration from dwm and bspwm.
+    '';
+    homepage = "https://github.com/ifreund/river";
+    license = licenses.gpl3Plus;
+    platforms = platforms.linux;
+    maintainers = with maintainers; [ branwright1 ];
```
Comment (fortuneteller2k) [inline pkgs/applications/window-managers/river/default.nix:37]: Add yourself to the [maintainer list](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix) with a separate commit `maintainers: add branwright1`

File: pkgs/applications/window-managers/river/default.nix [hunk 5]
```diff
@@ -0,0 +1,39 @@
+{ lib, stdenv ,fetchFromGitHub
+, zig, pkg-config, scdoc
+, wayland ,xwayland, wayland-protocols, wlroots
+, libxkbcommon, pixman, libudev, libevdev, libX11, libGL
+}:
+
+stdenv.mkDerivation rec {
+  name = "river";
```
Comment (fortuneteller2k) [inline pkgs/applications/window-managers/river/default.nix:8]:
```suggestion
  pname = "river";
  version = "unstable-2021-04-08";
```

File: pkgs/applications/window-managers/river/default.nix [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  installPhase = "zig build -Drelease-safe -Dxwayland -Dman-pages --prefix $out install";
+
+  nativeBuildInputs = [ zig scdoc pkg-config ];
+
+  installFlags = [ "DESTDIR=$(out)" ];
+
+  meta = with lib; {
+    description = "A dynamic tiling wayland compositor";
+    longDescription = ''
+    river is a dynamic tiling wayland compositor that takes inspiration from dwm and bspwm.
```
Comment (fortuneteller2k) [inline pkgs/applications/window-managers/river/default.nix:32]:
```suggestion
    river is a dynamic tiling wayland compositor that takes inspiration from dwm and bspwm.
```

**Review by moni-dz — CHANGES_REQUESTED** [review]

**Comment by branwright1** [comment]
Ok, changed river commit to unstable 2021-04-06 from master and added myself to maintainers.

**Comment by fortuneteller2k** [comment]
Add a `river` entry in the [top-level package list](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/all-packages.nix).

```nix
river = callPackage ../applications/window-managers/river { };
```

**Comment by branwright1** [comment]
Ofbord failed after adding callPackages to `all-packages.nix` because libudev is restricted for trusted users

**Comment by fortuneteller2k** [comment]
Result of `nixpkgs-review pr 119387` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>river</li>
  </ul>
</details>

File: pkgs/applications/window-managers/river/default.nix [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  src = fetchFromGitHub {
+    owner = "ifreund";
+    repo = "river";
+    rev = "9e3e92050e04320949c6cd995273c30319ebd515";
+    sha256 = "1v8dpbadsb3c7bc84sai09dbqv5s5s5d77vs12kdkd45x0ppmk3j";
+    fetchSubmodules = true;
+  };
+
+  buildInputs = [ wayland xwayland wayland-protocols wlroots pixman
+    libxkbcommon pixman libudev libevdev libX11 libGL
```
Comment (fortuneteller2k) [inline pkgs/applications/window-managers/river/default.nix:19]:
```suggestion
    libxkbcommon pixman udev libevdev libX11 libGL
```

File: pkgs/applications/window-managers/river/default.nix [hunk 8]
```diff
@@ -0,0 +1,39 @@
+{ lib, stdenv ,fetchFromGitHub
+, zig, pkg-config, scdoc
+, wayland ,xwayland, wayland-protocols, wlroots
+, libxkbcommon, pixman, libudev, libevdev, libX11, libGL
```
Comment (fortuneteller2k) [inline pkgs/applications/window-managers/river/default.nix:4]:
```suggestion
, libxkbcommon, pixman, udev, libevdev, libX11, libGL
```

**Review by moni-dz — CHANGES_REQUESTED** [review]

File: pkgs/applications/window-managers/river/default.nix (hunk 8 above)
Comment (fortuneteller2k) [reply pkgs/applications/window-managers/river/default.nix:4]: `libudev` is an alias for `udev`

File: pkgs/applications/window-managers/river/default.nix (hunk 7 above)
Comment (fortuneteller2k) [reply pkgs/applications/window-managers/river/default.nix:19]: `libudev` is an alias for `udev`

**Comment by branwright1** [comment]
How do I know which packages are aliasses?

**Comment by fortuneteller2k** [comment]
> How do I know which packages are aliasses?

Check this [file](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/aliases.nix)

**Review by moni-dz — APPROVED** [review]
Diff LGTM.

**Comment by fortuneteller2k** [comment]
Result of `nixpkgs-review pr 119387` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>river</li>
  </ul>
</details>

File: pkgs/applications/window-managers/river/default.nix [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+  src = fetchFromGitHub {
+    owner = "ifreund";
+    repo = "river";
+    rev = "9e3e92050e04320949c6cd995273c30319ebd515";
+    sha256 = "1v8dpbadsb3c7bc84sai09dbqv5s5s5d77vs12kdkd45x0ppmk3j";
+    fetchSubmodules = true;
+  };
+
+  buildInputs = [ wayland xwayland wayland-protocols wlroots pixman
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/river/default.nix:19]: Is a binary from wayland executed at build time? If so it needs to be in nativeBuildInouts.

File: pkgs/applications/window-managers/river/default.nix [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    libxkbcommon pixman udev libevdev libX11 libGL
+  ];
+
+  preBuild = "export HOME=$TMPDIR;";
+  installPhase = "zig build -Drelease-safe -Dxwayland -Dman-pages --prefix $out install";
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/river/default.nix:24]:
```suggestion
  preBuild = ''
    export HOME=$TMPDIR
  '';

  installPhase = ''
    zig build -Drelease-safe -Dxwayland -Dman-pages --prefix $out install
  '';
```

File: pkgs/applications/window-managers/river/default.nix [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+  nativeBuildInputs = [ zig scdoc pkg-config ];
+
+  installFlags = [ "DESTDIR=$(out)" ];
+
+  meta = with lib; {
+    description = "A dynamic tiling wayland compositor";
+    longDescription = ''
+    river is a dynamic tiling wayland compositor that takes inspiration from dwm and bspwm.
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/river/default.nix:33]:
```suggestion
      river is a dynamic tiling wayland compositor that takes inspiration from dwm and bspwm.
```

File: pkgs/applications/window-managers/river/default.nix (hunk 9 above)
Comment (branwright1) [reply pkgs/applications/window-managers/river/default.nix:19]: Wayland is xml file, there is no binary for it, I think it's runtime dependencie

**Comment by branwright1** [comment]
@SuperSandro2000 Can you check if it's ok now? commited and squashed those changes.

**Pull request merged**